agenthub/codeact_agent/codeact_agent.py

```diff
@@ -199,41 +196,83 @@ def step(self, state: State) -> Action:
         - MessageAction(content) - Message action to run (e.g. ask for clarification)
         - AgentFinishAction() - end the interaction
         """
-        messages: list[dict[str, str]] = [
-            {'role': 'system', 'content': self.system_message},
-            {'role': 'user', 'content': self.in_context_example},
-        ]
-
-        for prev_action, obs in state.history:
-            action_message = get_action_message(prev_action)
-            if action_message:
-                messages.append(action_message)
-
-            obs_message = get_observation_message(obs)
-            if obs_message:
-                messages.append(obs_message)
+        logger.info(f'Running CodeActAgent v{self.VERSION}')
+
+        # if we're done, go back
+        latest_user_message = state.history.get_latest_user_message()
+        if latest_user_message and latest_user_message.strip() == '/exit':
+            return AgentFinishAction()
+
+        # prepare what we want to send to the LLM
+        messages: list[dict[str, str]] = self._get_messages(state)
+
+        response = None
+        # give it multiple chances to get a response
+        # if it fails, we'll try to condense memory
+        attempt = 0
+        while not response and attempt < 10:
```

Comment:
Fixed. I'm not sure about the 2 though, because the PR implements [this approach](https://github.com/OpenDevin/OpenDevin/pull/1771#pullrequestreview-2065001341) of very lazy summary actions, where one `condense` round deals with one chunk. Then it immediately returns to try `completion` again. When the `response` to that is still unsuccessful, and it's still because of context limit or token limit, only then we try again to look for the next chunk to summarize.
As such, it seems to me perfectly possible for one to be unlucky enough to find 2 small chunks, too small to solve the context window limit. Then, if we stop after those 2 chunks, the only alternative is to completely fail the task...

That said, it seems clear to me that we don't need to get details like optimal number of attempts right off the start. IMHO all this algo will anyway see more iterations. 😅 And as requested, this is configurable now.

Reply:
That makes sense! I'll mark this as resolved.